### PR TITLE
Fix batched writes

### PR DESCRIPTION
**Pull request by otoolep**
This change involved a significant refactor, and two bug fixes. This change fixes issues #1499.

- Move code that populates individual points with any missing tags or timestamps from the top-level, to the influxdb package. This was done so this code is available for transports other than HTTP.
- Fix a panic, which occurred when the individual point had no tags, but the top-level object did. The panic occurred during the copy of the tags.
- Correctly detect when the timestamp key was not set on a individual point, thereby triggering the population of individual points' timestamps from the top-level.

### CODE REVIEW

**Comment by otoolep** [comment]
@corylanou @dgnorton

**Comment by otoolep** [comment]
This change also removed an unnecessary `for` loop, and sanity-checks incoming points.

File: influxdb.go [hunk 1]
```diff
@@ -109,6 +119,108 @@ var (
...
+	for _, p := range bp.Points {
+		if p.Timestamp.Time().IsZero() {
+			p.Timestamp = client.Timestamp(bp.Timestamp)
+		}
+		if p.Precision == "" && bp.Precision != "" {
+			p.Precision = bp.Precision
+		}
+		p.Timestamp = client.Timestamp(client.SetPrecision(p.Timestamp.Time(), p.Precision))
+		if len(bp.Tags) > 0 {
+			if p.Tags == nil {
```
Comment (otoolep) [inline influxdb.go]: This is the fix for the panic. If the individual point did not have tags, `p.Tags` was `nil`.

**Comment by otoolep** [comment]
Looking into test failures now.

File: httpd/handler_test.go [hunk 2]
```diff
@@ -87,13 +87,13 @@ func TestBatchWrite_UnmarshalEpoch(t *testing.T) {
 		t.Logf("testing %q\n", test.name)
```
Comment (corylanou) [inline httpd/handler_test.go]: This would be a good spot to add a quick test for sending in no timestamp and making sure it comes back as a zero time.

**Comment by corylanou** [comment]
LGTM.  Adding a quick test might be in order before closing this PR.  Feel free to merge on green.

File: httpd/handler_test.go (hunk 2 above)
Comment (otoolep) [inline httpd/handler_test.go]: Let me add that in a separate PR -- I will note it down now, as I have an idea about this testing that is larger in scope.

**Comment by otoolep** [comment]
Unit test code updated to enforce serial publishing by mock messaging client.

File: influxdb.go [hunk 3]
```diff
@@ -109,6 +118,108 @@ var (
...
+		bp.Points = epoch.Points
+		bp.Database = epoch.Database
+		bp.RetentionPolicy = epoch.RetentionPolicy
+		bp.Tags = epoch.Tags
+		bp.Timestamp = ts
+		bp.Precision = epoch.Precision
+		return nil
+	}(); err == nil {
+		return nil
+	}
```
Comment (benbjohnson) [inline influxdb.go]: Why is this section wrapped in an anonymous function?

**Comment by benbjohnson** [comment]
:+1:

File: influxdb.go (hunk 3 above)
Comment (corylanou) [inline influxdb.go]: Mostly because the conditional logic makes it indent really messy.  I could break it into a named function outside of this if we are against anonymous functions.

Comment (corylanou) [inline influxdb.go]: @benbjohnson once this is merged, I'll open a PR.  After looking at this, this really should be 2 more methods off of BatchPoint.  One for an `epoch` unmarshal, and one for a `normal` unmarshal.  That will be very clear, and make the top level `unmarshal` very short and readable.

Here is the issue I opened for it: https://github.com/influxdb/influxdb/issues/1507

**Comment by otoolep** [comment]
Thanks for the reviews.
